### ISSUE TITLE
Question about training restyle on a different dataset

### ISSUE THREAD

**Issue by DaddyWesker** [yuval-alaluf/restyle-encoder#8]
Hello again.

Recently, I've decided to train your Restyle on another dataset. Question is, do i need to train StyleGan or any other model your Restyle is using on the same dataset?

Thanks in advance for the answer.

**Comment by yuval-alaluf** [comment]
Hi @DaddyWesker,  
Yes. You should first train a StyleGAN generator on your domain. 
If your domain is one of the ones I uploaded to the repo (faces, cars, horses, churches, animal faces) you can probably using the pre-trained generator. If not, you'll need to train the generator. 
I recommend using one of the options described in the Readme. After training your generator you can plug in your generator and train the encoder using ReStyle.

**Comment by DaddyWesker** [comment]
So, if i got everything right. i need to 1. Train Stylegan on my dataset 
2. Use it's checkpoint to train Restyle, right?

And one more thing, how many pictures of one class is necessary to train properly both datasets? Any clues?

**Comment by yuval-alaluf** [comment]
Yep. I recommend option 2 that I [listed in the README](https://github.com/yuval-alaluf/restyle-encoder#preparing-your-generator) for training your StyleGAN since this will ensure you have everything you need for training ReStyle after. Just make sure that you save the average latent code since this is needed in ReStyle. 
Once you have the generator, the repo provides everything you will need for training (e.g., pre-trained backbone for the encoder and pre-trained networks for loss computations). All you need to do is define your custom dataset which I explain [here](https://github.com/yuval-alaluf/restyle-encoder#preparing-your-data).

Regarding the number of pictures necessary to train StyleGAN and ReStyle, it really depends. StyleGAN-ADA showed that you need far fewer samples than previous generators, but I am not sure of the recommended amount. I think the fewest number of samples I trained on was 5000. 

If you have any more questions feel free to ask.

**Comment by DaddyWesker** [comment]
Alright. Thanks for the explanations. I'll experiment a bit with training. Will close issue for now.

**Comment by DaddyWesker** [comment]
Hello again. I've been training stylegan2 until now and suspecting that i'm kinda messed it up. I've been training stylegan2-pytorch from this repo https://github.com/rosinality/stylegan2-pytorch not from official nvidia's repo (as i understand, that is exactly what i should've done). So, i've got ckpt from rosinality's stylegan2 and wanted to launch restyle training usng it, unfortunately i've got 
`Missing key(s) in state_dict: "convs.12.conv.weight", "convs.12.conv.blur.kernel", "convs.12.conv.modulation.weight", "convs.12.conv.modulation.bias", "convs.12.noise.weight", "convs.12.activate.bias", "convs.13.conv.weight", "convs.13.conv.modulation.weight", "convs.13.conv.modulation.bias", "convs.13.noise.weight", "convs.13.activate.bias", "convs.14.conv.weight", "convs.14.conv.blur.kernel", "convs.14.conv.modulation.weight", "convs.14.conv.modulation.bias", "convs.14.noise.weight", "convs.14.activate.bias", "convs.15.conv.weight", "convs.15.conv.modulation.weight", "convs.15.conv.modulation.bias", "convs.15.noise.weight", "convs.15.activate.bias", "to_rgbs.6.bias", "to_rgbs.6.upsample.kernel", "to_rgbs.6.conv.weight", "to_rgbs.6.conv.modulation.weight", "to_rgbs.6.conv.modulation.bias", "to_rgbs.7.bias", "to_rgbs.7.upsample.kernel", "to_rgbs.7.conv.weight", "to_rgbs.7.conv.modulation.weight", "to_rgbs.7.conv.modulation.bias", "noises.noise_13", "noises.noise_14", "noises.noise_15", "noises.noise_16". `

As i said previusly, i guess i messed this up a bit. So, any solution except re-training stylegan2 now from official nvidia's repo? That is really long thing to do actually...

**Comment by yuval-alaluf** [comment]
Hi @DaddyWesker,
I can related to the frustration of getting all the different components to work correctly. 
What size StyleGAN did you train? Did you set the output set of Restyle to match? 
Based on the error it seems like the StyleGAN the Restyle code is defining is larger than the one you trained (therefore its missing the last layers when trying to load the checkpoint).

**Comment by DaddyWesker** [comment]
Yes, this StyleGan i've trained is slightly smaller than one your code is expect to find. It has from conv 0 to conv 11. But, as i understand, your code expect to find layers till conv 15. 

I can't answer your question about "Did you set the output set of Restyle to match?". I've just launched it as it is, just prepared data using their "prepare_data" script.

**Comment by yuval-alaluf** [comment]
For training, there is a flag `--output_size` which by default is 1024, but you can set it to the resolution of the stylegan you trained (I am assuming 256)

**Comment by DaddyWesker** [comment]
Yes, i've used 256.

**Comment by DaddyWesker** [comment]
So, i guess, it had to be 1024?

**Comment by DaddyWesker** [comment]
FOrgot to mention, images from the dataset is 375 x 500 (H x W), so that's why i've created 256 x 256 from them

**Comment by yuval-alaluf** [comment]
Try setting the output_size flag to 256 when training Restyle

**Comment by DaddyWesker** [comment]
*facepalm* damn, i'm feeling myself kinda stupid right now. Yes, that worked. ANother problem is, that self.net.latent_avg is None (since if 'latent_avg' in ckpt: is False, as i understand, it should be in the loaded stylegan checkpoint), and because of that now i'm getting error here
```
self.avg_image = self.net(self.net.latent_avg.unsqueeze(0),
								  input_code=True,
								  randomize_noise=False,
								  return_latents=False,
								  average_code=True)[0]
```
SInce latent_avg is None it can't be unsqueezed. Can it be computed somehow? Or could something else used here instead of latent_avg?

**Comment by yuval-alaluf** [comment]
Yea. rosinality does not save the latent average unfortunately. One option is to randomly sample many latents and average them to see if they give you something reasonable. Another option is to simply start from the 0 latent vector. I didn't play with different initialization methods, but I believe different initializations will work just fine.

**Comment by DaddyWesker** [comment]
ALrighty, i'll try simplest way right now, thanks.
```
    def __load_latent_avg(self, ckpt, repeat=None):
        if 'latent_avg' in ckpt:
            self.latent_avg = ckpt['latent_avg'].to(self.opts.device)
            if repeat is not None:
                self.latent_avg = self.latent_avg.repeat(repeat, 1)
        else:
            self.latent_avg = torch.zeros(512).to(self.opts.device)
            # self.latent_avg = None
```
I've launched the training, though i was able to launch it only with batch_size = 4 (11 gb was eaten with batch 8). And one more thing, is it possible to launch your restyle training on several gpus? CUrrently it utilizes only one and i believe, it will be kinda too long to train. 

Btw, how long does it takes to train restyle_psp? What GPU you have used and for how long have you trained?

**Comment by yuval-alaluf** [comment]
The code as is doesn't support multiple GPUs. I trained all the encoders on P40 GPUs. If I remember I trained for a week to a week and a half depending on the StyleGAN size. I think for a 256 resolution generator training was only a couple of days.

**Comment by DaddyWesker** [comment]
ALrighty. Thanks for the answers. I'll see if i can do something with mGPU computing since i have only gtx 1080ti (though 4 of them) and it doesn't has 24 gb of vram to use big batch_size. And i'm not aware about performance difference between 1080 ti and p40. 
CLosing the issue for now
